File: 2020-06-month-long-challenge/day09.py
#!/usr/bin/env python3

# Day 9: Is Subsequence
#
# Given a string s and a string t, check if s is subsequence of t.
# A subsequence of a string is a new string which is formed from the original
# string by deleting some (can be none) of the characters without disturbing
# the relative positions of the remaining characters. (ie, "ace" is a
# subsequence of "abcde" while "aec" is not).
#
# Constraints:
# - 0 <= s.length <= 100
# - 0 <= t.length <= 10^4
# - Both strings consists only of lowercase characters.

class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        if len(s) == 0:
            return True
        if len(t) == 0:
            return False
        else:
            return s[0] in t \
                and self.isSubsequence(s[1:], t[t.index(s[0]) + 1:])

# Tests
assert Solution().isSubsequence("abc", "ahbgdc") == True
assert Solution().isSubsequence("axc", "ahbgdc") == False

File: 2020-06-month-long-challenge/day09.js
#!/usr/bin/env nodejs

/*
 * Day 9: Is Subsequence
 *
 * Given a string s and a string t, check if s is subsequence of t.
 * A subsequence of a string is a new string which is formed from the original
 * string by deleting some (can be none) of the characters without disturbing
 * the relative positions of the remaining characters. (ie, "ace" is a
 * subsequence of "abcde" while "aec" is not).
 *
 * Constraints:
 * - 0 <= s.length <= 100
 * - 0 <= t.length <= 10^4
 * - Both strings consists only of lowercase characters.
 */

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isSubsequence = function(s, t) {
    if (s.length == 0) {
        return true
    }
    if (t.length == 0) {
        return false
    }
    return t.indexOf(s[0]) != -1
        && isSubsequence(s.substring(1), t.substring(t.indexOf(s[0]) + 1 ))
};

// Tests
console.assert(isSubsequence("abc", "ahbgdc") == true)
console.assert(isSubsequence("axc", "ahbgdc") == false)
